### ISSUE TITLE
Feature request: Enable use of unix-style line endings when running sqlfluff fix command on Windows

### ISSUE THREAD

**Issue by nek555** [sqlfluff/sqlfluff#797]
In our environment, we use Windows PCs for development, but the company standard is that SQL files have unix-style line endings (LF).  We use bash under WSL a lot, and set Windows editors to use LF only.  However, the only python we have available is on Windows. When running sqlfluff fix under python on Windows, the file gets changed from LF to Windows-style line endings CRLF.

I understand that this is python default behavior, but it would be nice to be able to over-ride the default and choose LF only, all the time.

**Comment by NiallRees** [comment]
I think the answer to this is just a config option to specify line endings when writing files, with options of maybe `system default`, LF and CRLF?

**Comment by nek555** [comment]
A config option would be perfect.  For completeness, may as well include CR for Classic Mac, although probably not use much any more.

**Comment by jpy-git** [comment]
I'm going to take a look at adding this as I've got mac and windows users on my team and hate seeing those carriage returns :) 

@NiallRees's config option suggestion is definitely the right approach.

**Comment by jpy-git** [comment]
FYI the config option only approach was not appropriate on further investigation (see closed pr) but have a rule based approach will be more robust

**Comment by jpy-git** [comment]
@barrywhart @NiallRees update on where I've got with this rule:

1. Initially tried the config option approach for writing files, which worked fine (#1990). However, the issue with that is that we only write files if we fix at least one error, meaning that if a file had no issues other than inconsistent or wrong line endings we wouldn't re-export it and therefore the newlines don't get fixed. That's misleading for the user who would expect that sqlfluff would fix all the newlines so I closed that PR.

2. Then tried a rule based approach. In theory very simple, check all newline segments in the file and edit any that are inconsistent or don't match the config option. However, I ran into 2 issues: The first is that python imports files with universal newlines, which is easy enough to disable using `newline=''`. The second was that when Jinja renders the template (https://github.com/sqlfluff/sqlfluff/blob/4e21198f58b8784a7551e246708a7a6879baff09/src/sqlfluff/core/templaters/jinja.py#L343) it normalises the newlines automatically
![image](https://user-images.githubusercontent.com/80432516/146448055-0fe75dd0-d07e-45d6-a908-cc707e1e38d9.png)
and unfortunately it's not possible in Jinja to turn off this universal newline.
You can change the newline type it uses but can't keep whatever mix the user provides.
![image](https://user-images.githubusercontent.com/80432516/146448645-1a3b3cbd-a4f4-4900-bf6d-9567d22ed5df.png) (docs: https://jinja.palletsprojects.com/en/3.0.x/api/#newline_sequence)
This means it's not possible to maintain the different newline types into the parse tree and therefore perform rule analysis on them. I did try a workaround by mapping the newline segments back to their raw slices, but found that raw slices never matched perfectly with the original newline (would contain extra surrounding chars) and even when I could extract the true newline types from the raw I wouldn't be able to apply a fix because the parse tree would be unchanged and therefore the fix gets thrown away.

Because of this I think the most appropriate approach for this is to suggest that users instead consider the [mixed-line-ending](https://github.com/pre-commit/pre-commit-hooks/blob/master/pre_commit_hooks/mixed_line_ending.py) pre-commit hook or the `dos2unix` command, which they can just add to their pre-commit or CI. I've added a PR to ensure that all newlines supplied to SQLFluff are unix-normalised for internal analysis to ensure consistency, as is common with most python applications. PR: #2128

TLDR: don't think SQLFluff is the best place for newline conversion and there are other good tools for this as described above (UNIX philosophy and all).

@barrywhart would be good to get your opinion on the Jinja rendering part above, I assume it could be possible to patch those newlines back in but may not be simple given the potential that it could mess up the templating slicing. I'm of the opinion that it's not worth adding.

**Comment by barrywhart** [comment]
I'm good with this decision -- why jump through tricky hoops to solve a problem that already has a solution. In principle, I could imagine revisiting this someday, but we have more important challenges for the foreseeable future.

**Comment by nek555** [comment]
I appreciate that a large effort was made to try to fix this and it's 
too bad that this could not be done even with the large effort.

It leaves me with a less than ideal situation.  I work in an environment 
where I must use python on Windows, and I'm typically editing with 
Notepad++ and checking code in manually (not using CI/CD in our 
workstreams).  I'm well aware of dos2unix, but it's not available as a 
DOS command (AFAIK).  So, I have to manually change the line endings in 
Notepad++ and I have to do this (or at least check it) each time I use 
sqlfluff.

So, I hope there might someday be an enhancement for this.

Anyway sqlfluff is very useful to our team, thanks for all the efforts.

Norbert Kremer

**Comment by jpy-git** [comment]
@nek555 even if you don't want to use the linked pre-commit hook or dos2unix, you could very easily make a script to do your newline conversions in Python?

It would literally just be reading in the file with normalised unix-newlines and writing it out again:
```python
with open(file, mode="r", newline="\n") as f:
    contents = f.read()

with open(file, mode="w", newline="\n") as f:
    f.write(contents)
```
(you can wrap this in a CLI or whatever you need and would work fine on Windows)

I think even if we could do the rule based approach for this in SQLFluff, I think it's not really a linting error, linters for other languages (e.g. flake8) won't check it. It's not like someone using Windows has a choice about line endings so seems annoying to flag it. And as highlighted in point 1 we can't do the approach of writing out every single file in the hope of silently fixing newlines.
No need to shoehorn into SQLFluff given how easy it is to do this separately 😄

**Comment by nek555** [comment]
thanks for those great suggestions.  it looks like there are a number of 
decent work-arounds to mitigate this issue.
cheers,
Norbert
